### PR TITLE
debug: Add drgn script

### PR DESCRIPTION
**Pull request by lsgunth**
drgn can be a useful tool to dump some information in case of a hang in
md. Add as script which dumps some stats on active stripes and the
number of stripes in each state.

This requires a relatively recent version of drgn to be installed
(let's say 0.0.20 which isn't released yet).

There are instructions to install from source here:

  https://github.com/osandov/drgn

In the future, it may be installed with pip.

You can try running this on the next hang you get. It's a useful tool but usually I write something for purpose based on the bug.

### CODE REVIEW

File: debug/md.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    return Object(mddev.prog_, "struct r5conf", address=mddev.private)
+
+def find_hashed_stripes(conf):
+    PAGE_SIZE = 4096
+    NR_HASH = PAGE_SIZE // sizeof(conf.stripe_hashtbl[0])
+
+    stripes = []
+    for i in range(NR_HASH):
+        for s in hlist_for_each_entry('struct stripe_head',
```
Comment (dlsloan) [inline debug/md.py:41]: I assume my changes are about to break this?

Comment (lsgunth) [reply debug/md.py:41]: Eh, only if it hits one of your special objects.

File: debug/md.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (iomartin) [inline debug/md.py]: nit: empty line

File: debug/raid5_active.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import sys
+
+import md
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("disk", nargs="?", default="md0")
+    args = parser.parse_args()
+
+    mddev = md.find_mddev(prog)
```
Comment (iomartin) [inline debug/raid5_active.py]: prog is undefined?

**Review by dlsloan — APPROVED** [review]
Looks useful to me

File: debug/raid5_active.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env drgn
+
+import argparse
+import sys
+
+import md
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("disk", nargs="?", default="md0")
```
Comment (iomartin) [inline debug/raid5_active.py:10]: disk is not used

File: debug/md.py (hunk 1 above)
Comment (dlsloan) [reply debug/md.py:41]: They are all r5hash objects after my change though, most are just inside stripe_head containers.

File: debug/raid5_active.py (hunk 3 above)
Comment (lsgunth) [reply debug/raid5_active.py]: That's a drgn magic variable.

File: debug/md.py (hunk 1 above)
Comment (lsgunth) [reply debug/md.py:41]: Yeah, so for the most part it won't be broken, except when it tries to dereference an object it thinks is a stripe_head but isn't.

We can add support later if it's needed.

File: debug/raid5_active.py (hunk 3 above)
Comment (iomartin) [reply debug/raid5_active.py]: Ah ok

**Review by iomartin — APPROVED** [review]

File: debug/raid5_active.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+        state_map = {}
+        for s in stripes:
+            state_map.setdefault(hex(s.state), []).append(s)
+
+        for state, lst in state_map.items():
+            print(f"  -- State: {state} Count: {len(lst)}")
+
+    except md.MDException as e:
+        print (e)
```
Comment (iomartin) [inline debug/raid5_active.py]: Oh found one more nit... a space after print

**Pull request merged**